corehq/apps/enterprise/models.py

```diff
@@ -12,3 +20,44 @@ class EnterprisePermissions(models.Model):
         blank=True,
         default=list
     )
+
+    @classmethod
+    @quickcache(['domain'], timeout=7 * 24 * 60 * 60)
+    def get_by_domain(cls, domain):
+        """
+        Get or create the configuration associated with the given domain's account.
+        Note that the domain may be the source domain, one of the controlled domains,
+        or another domain in the account that does not use enterprise permissions.
+        """
+        account = BillingAccount.get_account_by_domain(domain)
+        try:
+            return cls.objects.get(account=account)
+        except cls.DoesNotExist:
+            return cls()
+
+    @classmethod
+    @quickcache(['source_domain'], timeout=7 * 24 * 60 * 60)
+    def get_domains(cls, source_domain):
+        """
+        Get a list of domains, if any, controlled by the given source domain.
+        """
+        try:
+            config = cls.objects.get(is_enabled=True, source_domain=source_domain)
+        except cls.DoesNotExist:
+            return []
+        return list(set(config.domains) - {config.source_domain})
+
+    def save(self, *args, **kwargs):
+        super().save(*args, **kwargs)
+        for domain in self.account.get_domains():
+            self.get_domains.clear(self.__class__, domain)
+            self.get_by_domain.clear(self.__class__, domain)
+
+    def clear_cache_for_all_users(self, domain=None):
+        from corehq.apps.domain.views.base import get_enterprise_links_for_dropdown
+        from corehq.apps.users.models import CouchUser
+        domains = [domain] if domain else self.account.get_domains()
+        for domain in domains:
+            for user_id in CouchUser.ids_by_domain(domain):
+                user = CouchUser.get_by_user_id(user_id)
+                get_enterprise_links_for_dropdown.clear(user)
```

Comment:
not loving the performance implications of this method but seems like it will only rarely be called. I do notice that it get's called during request processing, do you know how long this function will take? If it's longer than a few seconds it should be done asynchronously.

Reply:
I imagine it'll be bad on prod, I can pull it into an async task.

Reply:
07965ecfd59